firedrake/ml/pytorch.py

```diff
@@ -0,0 +1,210 @@
+from firedrake_citations import Citations
+
+from pyadjoint.reduced_functional import ReducedFunctional
+
+
+Citations().add("Bouziani2023", """
+@inproceedings{Bouziani2023,
+ title = {Physics-driven machine learning models coupling {PyTorch} and {Firedrake}},
+ author = {Bouziani, Nacime and Ham, David A.},
+ booktitle = {{ICLR} 2023 {Workshop} on {Physics} for {Machine} {Learning}},
+ year = {2023},
+ doi = {10.48550/arXiv.2303.06871}
+}
+""")
+
+
+__all__ = ['FiredrakeTorchOperator', 'torch_operator', 'to_torch', 'from_torch']
+
+
+class FiredrakeTorchOperator(torch.autograd.Function):
+    """PyTorch custom operator representing a set of Firedrake operations expressed as a reduced functional `F`.
+    `FiredrakeTorchOperator` is a wrapper around :class:`torch.autograd.Function` that executes forward and backward
+    passes by directly calling the reduced functional `F`.
+
+    Parameters
+    ----------
+    metadata : dict
+               Dictionary used to stash Firedrake objects.
+    x_P : list of torch.Tensor
+          PyTorch tensors representing the inputs to the Firedrake operator `F`.
+
+    Returns
+    -------
+    torch.Tensor
+          PyTorch tensor representing the output of the Firedrake operator `F`.
+    """
+
+    def __init__(self):
+        super(FiredrakeTorchOperator, self).__init__()
+
+    # This method is wrapped by something cancelling annotation (probably 'with torch.no_grad()')
+    @staticmethod
+    def forward(ctx, metadata, *x_P):
+        """Forward pass of the PyTorch custom operator.
+        """
+        F = metadata['F']
+        V = metadata['V_controls']
+        # Convert PyTorch input (i.e. controls) to Firedrake
+        x_F = [from_torch(xi, Vi) for xi, Vi in zip(x_P, V)]
+        # Forward operator: delegated to pyadjoint.ReducedFunctional which recomputes the blocks on the tape
+        y_F = F(x_F)
+        # Stash metadata to the PyTorch context
+        ctx.metadata.update(metadata)
+        # Convert Firedrake output to PyTorch
+        y_P = to_torch(y_F)
+        return y_P.detach()
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        """Backward pass of the PyTorch custom operator.
+        """
+        F = ctx.metadata['F']
+        V = ctx.metadata['V_output']
+        # Convert PyTorch gradient to Firedrake
+        adj_input = from_torch(grad_output, V)
+        if isinstance(adj_input, Function):
+            adj_input = adj_input.vector()
+
+        # Compute adjoint model of `F`: delegated to pyadjoint.ReducedFunctional
+        adj_output = F.derivative(adj_input=adj_input)
+
+        # Tuplify adjoint output
+        adj_output = (adj_output,) if not isinstance(adj_output, collections.abc.Sequence) else adj_output
+
+        # None is for metadata arg in `forward`
+        return None, *[to_torch(di) for di in adj_output]
+
+
+def torch_operator(F):
+    """Operator that converts a :class:`pyadjoint.ReducedFunctional` into a :class:`~FiredrakeTorchOperator`
```

Comment:
First line of docstring should be a one line description and shouldn't really have markup:
```python
"""Cast a Firedrake reduced functional to a PyTorch operator.

The resulting :class:`~FiredrakeTorchOperator` will take PyTorch tensors and inputs and return PyTorch tensors as outputs.
```